cmd/kubeless/function.go

```diff
@@ -100,14 +119,246 @@ func parseMemory(mem string) (resource.Quantity, error) {
 	return quantity, nil
 }
 
-func getFunctionDescription(funcName, ns, handler, funcContent, deps, runtime, topic, schedule, runtimeImage, mem string, triggerHTTP bool, envs, labels []string, defaultFunction spec.Function) (f *spec.Function, err error) {
+func getFileSha256(file string) (sha256Sum [32]byte, checksum string, err error) {
+	content, err := ioutil.ReadFile(file)
+	if err != nil {
+		return
+	}
+	h := sha256.New()
+	ff, err := os.Open(file)
+	if err != nil {
+		return
+	}
+	defer ff.Close()
+	_, err = io.Copy(h, ff)
+	if err != nil {
+		return
+	}
+	sha256Sum = sha256.Sum256(content)
+	checksum = hex.EncodeToString(h.Sum(nil))
+	return
+}
+
+type defaultPortForwarder struct {
+	cmdOut, cmdErr io.Writer
+}
+
+func (f *defaultPortForwarder) ForwardPorts(method string, url *url.URL, opts k8scmd.PortForwardOptions) error {
+	dialer, err := remotecommand.NewExecutor(opts.Config, method, url)
+	if err != nil {
+		return err
+	}
+	fw, err := portforward.New(dialer, opts.Ports, opts.StopChannel, opts.ReadyChannel, f.cmdOut, f.cmdErr)
+	if err != nil {
+		return err
+	}
+	return fw.ForwardPorts()
+}
+
+func getLocalPort() (string, error) {
+	for i := 30000; i < 65535; i++ {
+		conn, err := net.Dial("tcp", "127.0.0.1:"+strconv.Itoa(i))
+		if err != nil {
+			return strconv.Itoa(i), nil
+		}
+		conn.Close()
+	}
+	return "", errors.New("Can not find an unassigned port")
+}
+
+const (
+	maxRetries       = 5
+	defaultTimeSleep = 1
+)
+
+func waitForHTTPServer(host, port string) error {
+	// Build the request
+	req, err := http.NewRequest(
+		"GET",
+		"http://"+host+":"+port,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Execute the request
+	resp, err := http.DefaultClient.Do(req)
+	retries := 0
+	for err != nil && retries < maxRetries {
+		retries++
+		time.Sleep(time.Duration(time.Second))
+		resp, err = http.DefaultClient.Do(req)
+	}
+	resp.Body.Close()
+	return err
+}
+
+func runPortForward(f cmdutil.Factory, k8sClient *rest.RESTClient, podName, port string) (err error) {
+	clientset, err := f.ClientSet()
+	if err != nil {
+		return err
+	}
+	k8sClientConfig, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	portSlice := []string{port + ":9000"}
+	go func() {
+		devNull, err := os.Open(os.DevNull)
+		if err != nil {
+			logrus.Fatal(err)
+		}
+		pfo := k8scmd.PortForwardOptions{
+			RESTClient: k8sClient,
+			Namespace:  "kubeless",
+			Config:     k8sClientConfig,
+			PodName:    podName,
+			PodClient:  clientset.Core(),
+			Ports:      portSlice,
+			PortForwarder: &defaultPortForwarder{
+				cmdOut: devNull,
+				cmdErr: os.Stderr,
+			},
+			StopChannel:  make(chan struct{}, 1),
+			ReadyChannel: make(chan struct{}),
+		}
+		err = pfo.RunPortForward()
+		if err != nil {
+			logrus.Fatalf("Connection failed: %v", err)
+		}
+
+	}()
+	return
+}
+
+func getMinioURLFromOutsideCluster() (url string, err error) {
```

Comment:
As written, this function is un-unittestable, because it creates its own RESTClient without a chance to inject a fake one from outside.  As a coding pattern, we should _always_ pass in `kubernetes.Interface` or `rest.Interface` as parameters, and only the very top level `cmd/*.go` routines should actually create concrete instances of those classes.  See some of the existing unittests for how to create fake versions that are useful for tests.

(Ditto for other external ("real world") dependencies like file access, environment variables, etc)